CUSTOMCONFIG.md

```diff
@@ -0,0 +1,402 @@
+- Users can override these default values by loading a TOML file or passing values through CLI parameters.
+
+#### Main Function in `main.rs`:
+- In `main.rs`, the user-provided configuration (if any) is parsed. If a configuration file is provided, it is loaded into a `Config` object. If not, the `Config` object is instantiated with default values.
+- The `Config` object is then passed to the `Provision::new()` function, which sets up the system using the specified configuration.
+
+#### Example Flow in `main.rs`:
+```rust
+fn main() {
+    // Load the configuration file or use defaults
+    let config = load_config_from_file_or_default("/path/to/config.toml");
+    
+    // Initialize Provision with the Config object
+    let provision = Provision::new(config);
+
+    // Call the provision method to apply the configuration
+    provision.provision_ssh();
+    provision.provision_network();
+    // Other provisioning logic...
+}
+
+fn load_config_from_file_or_default(path: &str) -> Config {
+    // Attempt to load the config from a file, fallback to defaults if not found
+    match std::fs::read_to_string(path) {
+        Ok(content) => toml::from_str(&content).unwrap_or_default(),
+        Err(_) => Config::default(),
+    }
+}
+```
+
+### Pros of This Approach
+
+- **Single Source of Truth for Configuration**: The `Config` struct consolidates all configuration values, ensuring consistency and reducing the chance of duplication.
+
+- **Flexibility**: Users can either manually create a `Config` object (by specifying custom values) or load it from a file (e.g., TOML) or CLI arguments. This flexibility makes it easier to adapt the system for different environments or user needs.
+
+- **Standard Usage**: Many larger systems (such as `tokio` and `serde`) follow a similar pattern, encapsulating configuration in a struct that can be passed around. This design is well-understood and scalable, particularly in projects with complex configuration needs.
+
+### Cons of This Approach
+
+- **More Objects to Manage**: Users will need to manage both the `Config` and `Provision` objects. This might add some cognitive overhead, as they need to understand how to construct both.
+
+- **API Complexity**: This design might be more complex for users who are accustomed to a simple builder-style API. Instead of calling individual methods for each configuration option, they must work with the `Config` object, which is an API-breaking change.
+
+### Summary
+
+This approach centralizes configuration in a `Config` struct and passes it into `Provision`. It eliminates the need for a method-per-configuration approach and reduces duplication, providing a single source of truth for the configuration. Given the flexibility and consistency this provides, I believe this is the best approach for the long-term maintainability of the `azure-init` project.
+
+## Config Struct PR Changes and Notes - as of 10/25/2024
```

Comment:
We should have an example config with all the possible settings which represent azure-init out-of-box default behavior, or document that the azure-init.toml represents all the settings at their default value